utils/client.go

```diff
@@ -2,40 +2,82 @@ package utils
 
 import (
 	"context"
+	"encoding/json"
 	"log"
 
 	"github.com/gorilla/websocket"
-	"github.com/satori/go.uuid"
+	uuid "github.com/satori/go.uuid"
 
 	"github.com/spaceuptech/space-cloud/model"
+	pb "github.com/spaceuptech/space-cloud/proto"
 )
 
 // Client is the object handling all client interactions
 type Client struct {
-	id      string
-	channel chan interface{}
-	Context context.Context
-	cancel  context.CancelFunc
-	socket  *websocket.Conn
+	id       string
+	channel  chan *model.Message
+	Context  context.Context
+	cancel   context.CancelFunc
+	socket   *websocket.Conn              //only for Websocket
+	stream   pb.SpaceCloud_RealTimeServer //Only for grpc
+	protocol RealTimeProtocol
 }
 
 // DataCallback is the callback invoked when data is read by the socket
 type DataCallback func(data *model.Message)
 
 // RoutineWrite starts a json writer routine
 func (c *Client) RoutineWrite() {
-	go func() {
-		for res := range c.channel {
-			err := c.socket.WriteJSON(res)
-			if err != nil {
-				log.Println(err)
+	if c.protocol == Websocket {
+		go func() {
+			for res := range c.channel {
+				err := c.socket.WriteJSON(res)
+				if err != nil {
+					log.Println(err)
+				}
 			}
-		}
-	}()
+		}()
+	} else if c.protocol == GRPC {
+		go func() {
+			for res := range c.channel {
+				//Convert the Message into RealTime response.
+				switch res.Type {
+				case TypeRealtimeSubscribe, TypeRealtimeUnsubscribe:
+					//Decode the Message
+					var feedData []*pb.FeedData
+					responseMsg := res.Data.(model.RealtimeResponse)
+					for i, feed := range responseMsg.Docs {
+						payload, err := json.Marshal(feed.Payload)
+						if err != nil {
+							grpcResponse := pb.RealTimeResponse{Error: err.Error()}
+							c.stream.Send(&grpcResponse)
+						}
+						feedData[i] = &pb.FeedData{QueryId: feed.QueryID, DocId: feed.DocID, Type: feed.Type, Group: feed.Group, DbType: feed.DBType, Payload: payload, TimeStamp: feed.TimeStamp}
+					}
+					grpcResponse := pb.RealTimeResponse{Id: res.ID, Group: responseMsg.Group, Ack: responseMsg.Ack, Error: responseMsg.Error, FeedData: feedData}
+					c.stream.Send(&grpcResponse)
+
+				case TypeRealtimeFeed:
+					feed := res.Data.(model.FeedData)
+					var feedData []*pb.FeedData
+					payload, err := json.Marshal(feed.Payload)
+					if err != nil {
+						grpcResponse := pb.RealTimeResponse{Error: err.Error()}
+						c.stream.Send(&grpcResponse)
```

Comment:
Just log the error. Dont stream errors to the client. Also you need to return straight after logging the error.